Bugfix/adaptive selected actuators

seb5g opened the pull request:
should solve #609

seb5g committed:
handles errors when number of selected actuators and loss type is wrong or unknown

seb5g committed:
only set model on the *selected* actuators!

seb5g commented:
@malik-irain could you check it is now working?

malik-irain commented:
It seems to correct issue #609, but there are other important issues:

-  Before clicking "ini algo", if one is to select an actuator, then remove it before selecting another one, an exception is thrown, that also triggers another one  (`messagebox` isn't called correctly)
```python
Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/adaptive_optimization.py", line 131, in update_after_actuators_changed
    'lossdim').setValue(LossDim.get_enum_from_dim_as_int(len(actuators)))
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/loss_function/loss_factory.py", line 29, in get_enum_from_dim_as_int
    raise ValueError(f'No Loss with dim={dim} is known')
ValueError: No Loss with dim=0 is known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/optimizers_base/optimizer.py", line 609, in update_actuators
    self.update_after_actuators_changed(self.modules_manager.selected_actuators_name)
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/adaptive_optimization.py", line 134, in update_after_actuators_changed
    messagebox(f'You cannot select [{actuators}] as no corresponding Loss function exists')
  File "/home/mairain/original-pymodaq/pymodaq_gui/src/pymodaq_gui/messenger.py", line 21, in messagebox
    assert severity in MESSAGE_SEVERITIES
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError

```
- There are some problems with `runner_thread`, where `quit()` could be called before it exists, or after it was set to None (there's no need)
- When I change actuators between two runs, even when reinitializing the algorithm I get an exception:
```python
ERROR:pymodaq.optimizer:list index out of range
Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/optimizers_base/optimizer.py", line 186, in run_opti
    self.input_from_dets = self.model_class.convert_input(self.det_done_datas)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/optimizers_base/utils.py", line 285, in convert_input
    data_name: str = self.settings['optimizing_signal', 'optimize_0d']['selected'][0]
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

There could be other ones

seb5g commented:
ok thanks for the review, I will look into this more thoroughly!

seb5g committed:
proper use of messagebox

seb5g committed:
Update optimizer.py

seb5g committed:
create all prediction options at startup making them not visible

seb5g committed:
simplify the prediction function  initialization and robustness

seb5g committed:
pass correct settings for predictions

seb5g committed:
changing id names for prediction factory builders and adding convenience function to get usual names and short names

seb5g committed:
using short names for prediction id and update tradeoff

seb5g committed:
adding specific labels to have meaningfull plots

see https://github.com/PyMoDAQ/pymodaq_data/pull/46 for a patch on the subject

seb5g committed:
about the prediction function dynamically populate a 'options' settings with the content of the chossen prediction function arguments

seb5g commented:
A few patch to make sure the selected actuator appears correctly in the result plot (related to https://github.com/PyMoDAQ/pymodaq_data/pull/46 that should be accepted first). Also worked on the prediction function settings

seb5g commented:
@malik-irain could you check PR https://github.com/PyMoDAQ/pymodaq_data/pull/46 first then this one please?

malik-irain reviewed (changes requested):
I'll start with the most generic comments, that show area for improvment.

1. I find the flow to be a little bit unclear. The "Ini Model" button seems useless (because there is only one model right now, I know it's to be future-proof ;)), but it is not clear when we are supposed to selected devices to optimize for/with. It should be done before "Ini Algo", but doing so makes us click "Ini Model", select devices, click "Ini Algo". Even if the first two steps are inverted, we still need to go down then up on the window.

2. Selecting more than 2 actuators activate LossND, that doesn't offer any algorithm, but we can still click on "Ini Algo" and crash PyMoDAQ. It should be easily fixable as there is a more pressing issue related to that

3. Bounds are always [-5;5], if possible real bounds should be used?


Now, more concrete issues:
* changing algorithm while one is active shouldn't be allowed ("Ini Model", then "Ini Algo", then "Start", then "Pause", then change algorithm -> may crash), they sometimes need different parameters and may crash:
```python
	Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/adaptive_optimization.py", line 49, in queue_command
    self.optimization_algorithm.set_prediction_function(lossdim, kind, **command.attribute)
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/utils.py", line 43, in set_prediction_function
    self._prediction = LossFunctionFactory.create(loss_type, kind, **kwargs)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/loss_function/loss_factory.py", line 98, in create
    return cls.get(dim, key)()(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/loss_function/loss_factory.py", line 63, in __call__
    return self._loss(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^
TypeError: resolution_loss_function() got an unexpected keyword argument 'lower_threshold'
```

* when selecting "ini algo" we end up in `get_learner_from_enum` validating the condition `self == self.LOSS_1D` (because it's the default value), but bounds are empty. The "Ini Algo" button and toolbar button should be deactivated unless one or more actuator are selected.
```python
Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/optimizers_base/optimizer.py", line 683, in ini_optimization_runner
    self.set_algorithm()
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/adaptive_optimization.py", line 142, in set_algorithm
    self.algorithm = AdaptiveAlgorithm(
                     ^^^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/utils.py", line 37, in __init__
    self._algo = loss_type.get_learner_from_enum(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/loss_function/loss_factory.py", line 46, in get_learner_from_enum
    bounds = bounds[0]
             ~~~~~~^^^
IndexError: list index out of range

```
* In `optimizes_base/optimize.py::GenericOptimization.ini_optimization_runner`, `runner_thread.quit()` and following lines about thread termination may be called before thread is created. As it happens in combination with the previous issue, it might not be necessary to wrap it in a "if runner_thread:" or similar, but if done a log message saying a non-existing thread was asked to stop.
```python
Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/optimizers_base/optimizer.py", line 723, in ini_optimization_runner
    self.runner_thread.quit()
    ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'quit'
Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/optimizers_base/utils.py", line 263, in optimize_from
    self.modules_manager.get_det_data_list()
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/utils/managers/modules_manager.py", line 283, in get_det_data_list
    datas: DataToExport = self.grab_data()
                          ^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/utils/managers/modules_manager.py", line 340, in grab_data
    QThread.msleep(10)

```
* In Loss2D selecting triangle is impossible as it needs the `ip` argument that's not provided
```python
Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/optimizers_base/optimizer.py", line 683, in ini_optimization_runner
    self.set_algorithm()
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/adaptive_optimization.py", line 142, in set_algorithm
    self.algorithm = AdaptiveAlgorithm(
                     ^^^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/utils.py", line 39, in __init__
    loss_function=LossFunctionFactory.create(loss_type, kind, **kwargs))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/loss_function/loss_factory.py", line 98, in create
    return cls.get(dim, key)()(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/adaptive/loss_function/loss_factory.py", line 63, in __call__
    return self._loss(**kwargs)
           ^^^^^^^^^^^^^^^^^^^^
TypeError: triangle_loss() missing 1 required positional argument: 'ip'
```

So to sum up, "Ini Algo" should be activable only when there's 1 or 2 actuator (and relax the upper limit when LossND has algorithms). When "Ini Algo" is active, changing algorithm should be impossible. And disable or fix the Triangle loss function. It should make it good enough for normal usage!

seb5g committed:
better flow in selecting model/ini model...

if only one model, directly ini the model

seb5g committed:
add default actuator bounds

seb5g committed:
check if number of actuators is compatible with registered algo loss

malik-irain committed:
Config file is now from pymodaq and not pymodaq_utils

seb5g committed:
Merge branch 'bugfix/adaptive_sleected_actuators' of https://github.com/PyMoDAQ/PyMoDAQ into bugfix/adaptive_sleected_actuators

seb5g committed:
affect max data value positions as best individual

seb5g committed:
better algo flow handling

malik-irain commented:
A few problems remain:

- The triangle model is still not working (but could be disabled for the time being)
- It fails to run optimization when 2 actuator with different incompatible units are used (a real life example would be controlling a focal length in mm and a Zernike polynomial with no units ;)).

```python
ERROR:pymodaq.optimizer:Cannot convert from 'degree' (dimensionless) to 'degree_Celsius' ([temperature])
Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/optimizers_base/optimizer.py", line 208, in run_opti
    self.output_to_actuators.merge_as_dwa(
  File "/home/mairain/original-pymodaq/pymodaq_data/src/pymodaq_data/data.py", line 3096, in merge_as_dwa
    dwa.append(dwa_tmp)
  File "/home/mairain/original-pymodaq/pymodaq_data/src/pymodaq_data/data.py", line 1090, in append
    self.data += [Q_(data_array, data.units).m_as(self.units) for data_array in data.data]
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/.virtualenvs/original-pymodaq/lib/python3.12/site-packages/pint/facets/plain/quantity.py", line 315, in m_as
    return self.to(units).magnitude
           ^^^^^^^^^^^^^^
  File "/home/mairain/.virtualenvs/original-pymodaq/lib/python3.12/site-packages/pint/facets/plain/quantity.py", line 536, in to
    magnitude = self._convert_magnitude_not_inplace(other, *contexts, **ctx_kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/.virtualenvs/original-pymodaq/lib/python3.12/site-packages/pint/facets/plain/quantity.py", line 480, in _convert_magnitude_not_inplace
    return self._REGISTRY.convert(self._magnitude, self._units, other)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/.virtualenvs/original-pymodaq/lib/python3.12/site-packages/pint/facets/plain/registry.py", line 1050, in convert
    return self._convert(value, src, dst, inplace)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/.virtualenvs/original-pymodaq/lib/python3.12/site-packages/pint/facets/context/registry.py", line 405, in _convert
    return super()._convert(value, src, dst, inplace)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mairain/.virtualenvs/original-pymodaq/lib/python3.12/site-packages/pint/facets/nonmultiplicative/registry.py", line 267, in _convert
    raise DimensionalityError(src, dst, src_dim, dst_dim)
pint.errors.DimensionalityError: Cannot convert from 'degree' (dimensionless) to 'degree_Celsius' ([temperature])

```

seb5g committed:
correct for the fact that different actuaotrs may have different units hence not mergeable into a single DWA

seb5g committed:
adding an abstract method into models o check if a fitness value has been set in the model

seb5g committed:
enable run given the model/algo status

seb5g committed:
has to reorder the  suggested actuaotr value is teh selected actuators are not in sorted order...which is the default for bayesian algo space paramter

seb5g committed:
changing all methods and class to use an explicit fom of actuators coordinates

in the dorm of a dict with keys as names of the actuaotrs

seb5g committed:
using ordered dict for setting  bounds

seb5g committed:
bayesian seems to be working: todo check stopping and adapt adaptive

seb5g committed:
adding a restart optimization action

seb5g committed:
adapt adpative interface to new layout (better management)

seb5g committed:
patch for 1D adaptive learner

seb5g committed:
better handlgin of stop and restart

seb5g committed:
Merge branch '5.1.x_dev' into bugfix/adaptive_sleected_actuators

seb5g commented:
these have tremendous changes to make sure the optimization applies on the right actuators. 

I also introduced a restart button and better handling of what setting it is possible to change at each part of the optimization flow!

@malik-irain could you have a go for  another review

seb5g commented:
remark: test are failing right now because I just published a pymodaq_gui release, this version need... should restart them in a minute...

seb5g committed:
Update utils.py

seb5g committed:
Merge branch 'bugfix/adaptive_sleected_actuators' of https://github.com/PyMoDAQ/PyMoDAQ into bugfix/adaptive_sleected_actuators

seb5g committed:
remove triangle loss as it is extremely slow

seb5g commented:
finally I removed the triangle loss. While it was now working, it was extremely slow (as said in the doc...). Ready to merge this huge improvement!!!!!!!!!!!!!!

malik-irain commented:
Two points:

- The button to move to the best guess does not work (tested with a real setup), the actuator doesn't move at all
- The default model seems to target the points where the value is minimum. It's not especially an issue, but I just wanted to check if that was wanted as the others seem to targer points where the value is maximized

seb5g commented:
> Two points:
> 
>     * The button to move to the best guess does not work (tested with a real setup), the actuator doesn't move at all
> 
>     * The default model seems to target the points where the value is minimum. It's not especially an issue, but I just wanted to check if that was wanted as the others seem to targer points where the value is maximized

do you mean with adaptive or with bayesian?

malik-irain commented:
With adaptive

seb5g commented:
adaptive is an algotithm aiming at sampling non uniform distribution. I will try to dowsample regions where nothing much happens and upsample regions where the signal varies rapidly (either around a max or a min) so it makes sense I guess. Which function did you sample in your test?

seb5g committed:
create new goto action: if activated, double licking the plot area will move the actuaotrs

seb5g committed:
setting correctly the saving mechanisms

seb5g committed:
add code when user clicks on save action

seb5g committed:
introducing a StrEnum for the list of possible Actions

seb5g committed:
reload the runner if needed after a dosave

seb5g commented:
last commits solve your raised points and adds some further stability in the flow, better saving

seb5g committed:
adding status bar and stop button

seb5g commented:
@malik-irain I added the status bar and a stop button that stop momentary the algo and go to the best position also emitting a stop signal (its a manual algo stop convergence)

malik-irain commented:
I just discovered a new bug:

```
ERROR:pymodaq.optimizer:Data point [-10. -10.] is not unique. You can set "allow_duplicate_points=True" to avoid this error
Traceback (most recent call last):
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/optimizers_base/optimizer.py", line 223, in run_opti
    self.optimization_algorithm.tell(self.input_from_dets)
  File "/home/mairain/original-pymodaq/pymodaq/src/pymodaq/extensions/bayesian/utils.py", line 93, in tell
    self._algo.register(params=self._next_point, target=function_value)
  File "/home/mairain/.virtualenvs/original-pymodaq/lib/python3.12/site-packages/bayes_opt/bayesian_optimization.py", line 200, in register
    self._space.register(params, target, constraint_value)
  File "/home/mairain/.virtualenvs/original-pymodaq/lib/python3.12/site-packages/bayes_opt/target_space.py", line 477, in register
    raise NotUniqueError(error_msg)
bayes_opt.exception.NotUniqueError: Data point [-10. -10.] is not unique. You can set "allow_duplicate_points=True" to avoid this error

```

which should be easily fixable as the fix is given.

Once done, I'll accept the PR, but don't forget to document it!

seb5g commented:
what did you do to induce this? It seems to be on the bayes_opt side? I never encountered it?

malik-irain commented:
Nothing special, my supposition is the same point got probed twice with different values as I was using mock devices

seb5g committed:
factorising doc about extensions and adding adaptive and datamixer extrnsions doc

seb5g committed:
improving bayesian documentation

seb5g committed:
patch for the stopping criteria

adding absolute value to compare to tolerance
adding None as  a stopping criterium

seb5g committed:
adding the best stopping criteria

seb5g committed:
require pymodaq_utils 5.0.10

seb5g committed:
Update pyproject.toml

seb5g committed:
bayesian documentation is updated!

seb5g committed:
adaptive documentation

seb5g committed:
adding some bits to the doc

seb5g committed:
updated doc
